google/cloud/pubsublite/internal/service_composite.h

```diff
@@ -0,0 +1,159 @@
+    if (!shutdown.is_ready()) {
+      GCP_LOG(WARNING)
+          << "`Shutdown` must be called and finished before object "
+             "goes out of scope if `Start` was called.";
+      assert(false);
+    }
+    shutdown.get();
+  }
+
+  future<Status> Start() override {
+    future<Status> start_future;
+    std::vector<future<Status>> dependency_futures;
+
+    {
+      std::lock_guard<std::mutex> g{mu_};
+      for (auto* dependency : dependencies_) {
+        dependency_futures.push_back(dependency->Start());
+      }
+      start_future = status_promise_->get_future();
+      shutdown_ = false;
+      status_ = Status{};
+    }
+
+    for (auto& dependency_future : dependency_futures) {
+      dependency_future.then([this](future<Status> status_future) {
+        Status s = status_future.get();
+        if (!s.ok()) Abort(std::move(s));
+      });
+    }
+
+    return start_future;
+  }
+
+  /**
+   * This will add a `Service` dependency for the current object to manage. It
+   * is only added if the current object hasn't been `Shutdown` yet. May only be
+   * called after `Start`.
+   * @param dependency
+   */
+  void AddServiceObject(Service* dependency) {
+    future<Status> start_future;
+    {
+      // under lock to guarantee atomicity of being added to `dependencies_` and
+      // `Start` being called so `Start` called on dependency if and only if
+      // `Shutdown` will be called on dependency
+      std::lock_guard<std::mutex> g{mu_};
+      if (shutdown_) return;
+      dependencies_.push_back(dependency);
+      start_future = dependency->Start();
+    }
+    start_future.then([this](future<Status> status_future) {
+      Status s = status_future.get();
+      if (!s.ok()) Abort(std::move(s));
+    });
+  }
+
+  /**
+   * @note Can be safely called more than once.
+   */
+  void Abort(Status s) {
+    assert(!s.ok());
+    promise<Status> start_promise{null_promise_t{}};
+    {
+      std::lock_guard<std::mutex> g{mu_};
+      if (!status_promise_) return;
+      start_promise = std::move(*status_promise_);
+      status_promise_.reset();
+      status_ = s;
+    }
+    start_promise.set_value(std::move(s));
+  }
+
+  /**
+   *
+   * @return a `Status` of `kOk` if and only if the current lifecycle is in the
+   * running phase
+   */
+  Status status() {
+    std::lock_guard<std::mutex> g{mu_};
+    return status_;
+  }
+
+  future<void> Shutdown() override {
+    absl::optional<promise<Status>> status_promise;
+    {
+      std::lock_guard<std::mutex> g{mu_};
+      if (shutdown_) return make_ready_future();
+      shutdown_ = true;
+      status_promise_.swap(status_promise);
+      if (status_.ok())
+        status_ = Status{StatusCode::kAborted, "`Shutdown` called"};
+    }
+    if (status_promise) status_promise->set_value(Status());
+
+    AsyncRoot root;
+    future<void> root_future = root.get_future();
+
+    std::lock_guard<std::mutex> g{mu_};
+    for (auto* dependency : dependencies_) {
```

Comment:
I had a question here, but I think it was answered, I missed the effect of `ChainFuture()`.